partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt

```diff
@@ -396,8 +407,13 @@ data class ClobType(override val metas: Map<String, Any> = mapOf()) : SingleType
  */
 data class ListType(
     override val elementType: StaticType = ANY,
-    override val metas: Map<String, Any> = mapOf()
+    override val metas: Map<String, Any> = mapOf(),
+    override val constraints: Set<CollectionConstraint> = setOf()
 ) : CollectionType() {
+
+    init {
+        validateCollectionConstraints(elementType, constraints)
+    }
```

Comment:
For this and where else applicable: 

In a typical DBMS, most of the "SQL constraints", i.e., `primary key`, `unique`, `check`, etc  are validated during insertion/deletion.

Query execution, on the other hand, only scan the underlying data, it does not recheck the constraints. 

Do we planing on using the same concept? 

Assuming at the moment `validateCollectionConstraints` is just a place holder, and in the future you planning on doing the actual validation, i.e, uniqueness, etc. 

Consider we have a query `SELECT a FROM tbl1`, where `tbl1` is an external data source with `a` being the primary key. Does the from source gets a `schemaType` which includes the original primary key constraint or just a "plain collection" type with no primary key constraints. 

If former, then seemingly when we are loading the data, we also needs to validate the primary key constraints, which can be super time consuming. 

If later, then how are we going to retrieve the information that `a` is primary key, which at least means `a` is `unique` and `non-null`. Such information may comes in handy when conducting an plan optimization pass.

Reply:
> Assuming at the moment validateCollectionConstraints is just a place holder, and in the future you are planning on doing the actual validation, i.e, uniqueness, etc.

`validateCollectionConstraints` purpose is just to ensure we model the correct constraints on `Collection` when creating such types; at this stage, I don't see validating the actual data against this shape in the long run. On the other hand, the modelling of `SQL` constraints doesn't necessarily mean that we need to support the same physical operations on them; the main purpose it to be able to capture the shape of the data in various models incl. Relational and `SQL` correctly. The correct representation of the shape can be of many uses; for example—as you also pointed out—knowing the details of the schema can help us with planning optimization such as using other operators rather than the full table scan in the presence of primary key in `WHERE`  clause.

> Consider we have a query SELECT a FROM tbl1, where tbl1 is an external data source with a being the primary key. Does the from source gets a schemaType which includes the original primary key constraint or just a "plain collection" type with no primary key constraints.

In my mental model `tbl1` will have a type in `Γ0` which can very well be `ANY` in case of absence of schema; we would also have `tbl1` in `ρ0` which includes the data for the current round of evaluation; and then the constraint enforcement will be the concern of our type inference and checking algorithm. Having said that, back to my comment above, enforcing some schema constraints like `PrimaryKey` beyond planning and physically, may not be something that we want to be involved at all—as they can be an an iteration to our logical plan implemented by external query and storage engines.